### PR TITLE
Allow Webroot Certificate Creation

### LINKED ISSUE #39: Add certbot_create_method 'webroot' for zero-downtime cert generation
After the work in #12 and #38, this role supports using Certbot's `--standalone` server for generating certs, which is great and perfectly adequate if you're either:

  1. Generating the cert on the first bootstrap of a server (since it's not causing an outage, it's just prepping for the first webserver startup).
  2. Okay with a few seconds of webserver downtime when you generate a new cert (note that there's no downtime during renewals).

However, some of the examples in #12's comments showed how certbot's `--webroot` option might be used with either Apache or Nginx to allow running webservers to not have to be stopped before a cert is generated.

Examples:

  - https://github.com/geerlingguy/ansible-role-certbot/issues/12#issuecomment-302216046
  - https://github.com/geerlingguy/ansible-role-certbot/issues/12#issuecomment-264167972 (specific to Apache)

See docs: https://certbot.eff.org/docs/using.html#webroot

### LINKED ISSUE #41: Support deployhook
Add deployhook support

### LINKED ISSUE #42: Add test cert support
add support for using non prod CA

### PR DESCRIPTION
**Pull request by simonspa**
This PR supersedes #43 by @rfaircloth-splunk and adds support for the `webroot` generation method as well as requesting HSTS or running in test mode.

Fixes #41 support deployhook (webroot only)
Fixes #42 support for test CA
Fixes #39 Support for webroot

### CODE REVIEW

File: defaults/main.yml
Comment (kierenevans) [inline defaults/main.yml]: This doesn't work with multiple services (or even one service where it prints `service [u'nginx'] restart` instead.
Try this instead:
```suggestion
certbot_deployhook: "service {{ certbot_create_standalone_stop_services | join(' restart && service') }} restart"
```

**Comment by simonspa** [comment]
You're right @kierenevans thanks for the fix!

**Comment by MicroJoe** [comment]
Thanks for this PR, I am currently using it (rebased it on master) until not merged.

One little issue I have had is that the `/etc/letsencrypt/` directory does not seem to exist right after installing the `certbot` package on Debian 10 Buster, probably because `certbot.service` was not started right after installation (but it should have been, by whom, this role or Debian?).

```
TASK [geerlingguy.certbot : Check if certificate already exists.] **********************************************************************************************
ok: [***]

TASK [geerlingguy.certbot : Create deploy hook] ****************************************************************************************************************
fatal: [***]: FAILED! => {"changed": false, "checksum": "7fd17a05d6d10769b284884b86040525a5894358", "msg": "Destination directory /etc/letsencrypt/renewal-hooks/deploy does not exist"}
```

Manually starting `certbot.service` lead to the creation of the directories, which let the role go fine after.

File: defaults/main.yml
Comment (MicroJoe) [inline defaults/main.yml:32]: `cert_item.webroot | default(certbot_webroot)` could be useful in order to avoid to repeat `webroot` for each domain if it is common to all of them (in my case: `/var/www/letsencrypt`, not sure if this could be a good default value).

File: tasks/create-cert-webroot.yml
Comment (MicroJoe) [inline tasks/create-cert-webroot.yml:17]: Same comment, see https://github.com/MicroJoe/ansible-role-certbot/commit/4a114225885b03ab8c1e865c7409f49953c62396

File: defaults/main.yml
Comment (simonspa) [reply defaults/main.yml]: Very good idea!

**Comment by simonspa** [comment]
Hi @MicroJoe, Thanks for the suggestion, very good idea. I implemented it and it seems to work for me - care to test?

I also added checks to ensure the config directory actually exists - doesn't hurt probably.

**Comment by MicroJoe** [comment]
Thanks, it looks great!
I am currently provisioning a new cloud machine to test that the role works on clean machine now with the config directory verification.

**Comment by MicroJoe** [comment]
After testing, I think your last commit is not fixing the issue I encounter.
However, here is what fixed it for me: verify that the hook directory exist before copying it.

```diff
diff --git a/tasks/create-cert-webroot.yml b/tasks/create-cert-webroot.yml
index 5d0ac6d..63466c8 100644
--- a/tasks/create-cert-webroot.yml
+++ b/tasks/create-cert-webroot.yml
@@ -4,6 +4,12 @@
     path: /etc/letsencrypt/live/{{ cert_item.domains | first }}/cert.pem
   register: letsencrypt_cert
 
+- name: Ensure deploy hook directory exists
+  file:
+    path: /etc/letsencrypt/renewal-hooks/deploy
+    state: directory
+    mode: 0755
+
 - name: Create deploy hook
   copy:
     content: "{{ certbot_deployhook }}"
```

It should also be added to `create-cert-standalone.yml` for feature parity, but you get the idea.
As for your last commit, I am not sure if it is useful as this change will supersede your verification, so maybe you can adapt it. The less tasks the faster deployment. ;)

**Comment by simonspa** [comment]
Great, thanks for testing - I think I misread the error log you posted earlier...

I removed the unnecessary check again and added your patch. It's not necessary to add anything for standalone creation, there no deployment hook is written.

**Comment by MicroJoe** [comment]
Thanks, I have rebased my fork's `webroot` branch to match yours. I have nothing to add to this PR, I find it useful and hope it can be merged.

@geerlingguy any thoughts?

**Comment by MicroJoe** [comment]
After playing with this feature a little bit, I have noticed that the renewal-hook installation will occur once per domain, while there is nothing depending on the domains in the renewal-hook creation. This is slowing down the certificate generation loop as it is executed each time.

In https://github.com/MicroJoe/ansible-role-certbot/commit/c841d79d93e96bc7bf5467ded0dd16d1e6490a0e I patch your PR to not do the hook installation in the loop but outside of it. It was done in a hurry, but it can be reworked to go nice with this PR's content.

**Comment by simonspa** [comment]
Thanks, @MicroJoe - that was indeed unnecessary. I cherry-picked your commit onto this branch!

**Comment by MicroJoe** [comment]
When using this patch on a new machine I have had to change the hook creation to happen before the very first certificate is issued. This fix is visible in this commit: https://github.com/MicroJoe/ansible-role-certbot/commit/2346cd52a262bf2fa89419f29abba0e8ea95cdb7

**Comment by simonspa** [comment]
Thank you very much for reporting, @MicroJoe - I have cherry-picked your commit and added it to this branch.

**Comment by simonspa** [comment]
@geerlingguy any chance for you to review this and get it in? It is already used by a fair amount of people and has been reviewed by two independent parties. I believe that the functionality this adds to your role is worth the few minutes of review.

**Comment by simonspa** [comment]
It's inactive since it's just sitting here and waiting for a merge. I still think it's a feature worth having.

**Comment by geerlingguy** [comment]
I still like the feature too, and if you can rebase Lint should pass.

**Comment by simonspa** [comment]
Hi @geerlingguy 

sure, no problem, I rebased - but the Linter job fails, as it does on `master`. I have fixed a few things (fixing the line length in your github workflow, installing ansible in addition to ansible-lint` but I don't feel comfortable venturing deeper into fixing the ansible-lint errors in this merge request here - they are unrelated to the code discussed here.

Please check and let me know if I should do anything else.

/Simon

**Comment by geerlingguy** [comment]
Oh shoot... I accidentally committed the lint fixes on the wrong branch. I'll get that sorted momentarily.

**Comment by geerlingguy** [comment]
All right, please try rebasing again.

**Comment by simonspa** [comment]
Done, and linter passes :tada:

**Comment by simonspa** [comment]
@geerlingguy I implemented all requested changes - any plan to merge this?

File: defaults/main.yml
Comment (karlmdavis) [inline defaults/main.yml:57]: @simonspa Might I suggest this minor improvement to this excellent PR?

```suggestion
certbot_deployhook: |                                                                                                                                                                                     
  {% for service in certbot_create_standalone_stop_services %}
  if systemctl list-unit-files | grep -q "^{{ service }}.service"; then
    systemctl restart {{ service }}
  fi
  {% endfor %}
```

I often have a chicken-egg problem where the services I want to create certs for don't/can't exist until after certbot has run. Adding these `if`s avoids that problem by only trying to restart the services if they exist.

Comment (simonspa) [reply defaults/main.yml:57]: Thanks, LGTM!

File: defaults/main.yml
Comment (geerlingguy) [inline defaults/main.yml:36]:
```suggestion
  {{ cert_item.webroot | default(certbot_webroot) if certbot_create_method == 'webroot'  else '' }}
```

  {{ cert_item.webroot | default(certbot_webroot) if certbot_create_method == 'webroot' else '' }}

File: defaults/main.yml
Comment (geerlingguy) [inline defaults/main.yml:56]:
```suggestion
# Restart this service after successful certificate creation:
```

File: .github/workflows/release.yml
Comment (geerlingguy) [inline .github/workflows/release.yml:38]: Can you revert this change, only because it's unrelated and is standardized across all 60+ of my Galaxy roles.

**Review by geerlingguy — CHANGES_REQUESTED** [review]

File: .github/workflows/release.yml
Comment (simonspa) [reply .github/workflows/release.yml]: Done. That was the linter though...

**Comment by hkcomori** [comment]
In master, arguments have been added to certbot_create_command.
In webroot, I have errors that these directories are not found when first installation.

https://github.com/geerlingguy/ansible-role-certbot/pull/80/files#diff-00781efe962bc7f1510921c6d0bf5340fccdd1436a7507f76e766a1876a1fb03

**Comment by simonspa** [comment]
> In master, arguments have been added to certbot_create_command. In webroot, I have errors that these directories are not found when first installation.
> 
> https://github.com/geerlingguy/ansible-role-certbot/pull/80/files#diff-00781efe962bc7f1510921c6d0bf5340fccdd1436a7507f76e766a1876a1fb03

Thanks for the ping, I have rebased this branch on top of latest `master`. @geerlingguy All your comments have been addressed some five months ago, would you mind merging this? :)

/S

**Comment by geerlingguy** [comment]
Released in tag 5.0.0—please test and make sure things are working for you!

**Comment by simonspa** [comment]
Thanks, working flawlessly over here!

**Pull request merged**
